### PR TITLE
Feature/group management actions

### PR DESCRIPTION
**Pull request by Atilla13**
## Overview
Implements Leave Group, Share Group, and Delete Group actions with ownership validation, confirmation dialogs, and comprehensive test coverage.

### UI Layer (GroupListScreen.kt)
- Add confirmation dialogs for Leave and Delete actions
Screenshot 2025-11-05 at 12 21 05 AM

- Add restriction dialogs for ownership validation (owner can't leave, non-owner can't delete)
Screenshot 2025-11-05 at 12 21 32 AM

- Wire up share intent with system share sheet
Screenshot 2025-11-05 at 12 21 55 AM

### Repository Layer (GroupRepositoryFirestore.kt)
- Implement `deleteGroup()` and `leaveGroup()` with Firebase Auth validation
- Add exception handling for unauthorized actions

### ViewModel Layer (GroupListViewModel.kt)
- Add `deleteGroup()` and `leaveGroup()` methods with callbacks
- Refresh group list after operations

### Navigation (MainActivity.kt)
- Wire up callbacks for leave, delete, and share actions

## Test Coverage
- **Instrumentation**: Dialog flows, button interactions, callback verification
- **Unit**: Success/error handling, state management (12 tests)
- **Repository**: Ownership validation, edge cases (6 tests)
- Added `leaveGroup()` to FakeGroupRepository in 6 test files

### CODE REVIEW

File: app/src/main/java/com/android/joinme/model/groups/GroupRepositoryFirestore.kt [hunk 1]
```diff
@@ -50,9 +50,31 @@ class GroupRepositoryFirestore(private val db: FirebaseFirestore) : GroupReposit
   }
 
   override suspend fun deleteGroup(groupId: String) {
+    val currentUserId =
```
Comment (vlonfat) [inline app/src/main/java/com/android/joinme/model/groups/GroupRepositoryFirestore.kt]: Why fetch the userID? In leaveGroup, it is already fetched in the view model as it is passed as an argument, whereas in leaveGroup it is refetched. I propose two solutions:

-	Option 1: both functions take userID as an argument (because we fetch the userID in the view model but refetch it in delete group)
-  Option 2: create a global variable in the firestore userID and remove the userID argument from leave group. And therefore no longer fetch the userID in the viewModel. This option follows the same logic as for events and series.

File: app/src/main/java/com/android/joinme/MainActivity.kt [hunk 2]
```diff
@@ -262,15 +267,42 @@ fun JoinMe(
             onProfileClick = { navigationActions.navigateTo(Screen.Profile) },
             onEditClick = { navigationActions.navigateTo(Screen.EditProfile) },
             onViewGroupDetails = { navigationActions.navigateTo(Screen.GroupDetail(it.id)) },
-            onLeaveGroup = {
-              Toast.makeText(context, "Not yet implemented ", Toast.LENGTH_SHORT).show()
+            onLeaveGroup = { group ->
+              val currentUserId = FirebaseAuth.getInstance().currentUser?.uid
```
Comment (vlonfat) [inline app/src/main/java/com/android/joinme/MainActivity.kt]: This is something that should not appear in the composable! It must be handled in the viewModel for the userID!!

**Review by vlonfat — CHANGES_REQUESTED** [review]
# Summary
Good work but some details need to be review to ensure consistency, security and testing.

## Coverage Issues

Target: 80% line coverage required
Current status:
- GroupListScreen: 63% line coverage
- MainActivity: 0% line coverage
It would be preferable to set up this line coverage, but in an intelligent way so as not to add dozens and dozens of tests, otherwise the CI execution time will become very long.

## Architecture Inconsistency: deleteGroup vs leaveGroup
There's an inconsistency in how userId is handled between these two functions:

- deleteGroup receives userId as a parameter
- leaveGroup fetches userId internally

Proposed solutions:
**Option 1 (Recommended)**: Both functions take userId as a parameter
Avoids redundant fetching (currently fetched in ViewModel, then re-fetched in deleteGroup) and maintains consistency across functions

**Option 2:** Create a global userId variable in Firestore
Remove userId parameter from leaveGroup and remove userId fetching from ViewModel

## Code Quality Issue
userId is currently being fetched in the UI composable, which violates separation of concerns. This logic should be moved to the ViewModel. (file MainActivity)

## Design 
It is best to follow the figma directly when creating the UI. This saves time later on and ensures that features are clean when merged. Please update the UI directly as shown in the figma.

File: app/src/main/java/com/android/joinme/MainActivity.kt (hunk 2 above)
Comment (Atilla13) [reply app/src/main/java/com/android/joinme/MainActivity.kt]: Correct, thank you for pointing this out! I will definitely fix it!

**Comment by Atilla13** [comment]
To address your point:

> It is best to follow the Figma directly when creating the UI. This saves time later on and ensures that features are clean when merged. Please update the UI directly as shown in the Figma.

I believe it’s more efficient if I handle this as part of my Color Standardization & Dark Mode Foundation across all pages, ensuring the entire application aligns with the Figma design consistently at our current checkpoint.

Now I just need to boost MainActivity’s line coverage, and then everything should be wrapped up!  

Thanks a lot for your feedback.

File: app/src/androidTest/java/com/android/joinme/ui/navigation/MainActivityTest.kt [hunk 3]
```diff
@@ -90,4 +94,294 @@ class MainActivityTest {
...
+        assert(pathSegments?.firstOrNull() == "group")
+        assert(pathSegments?.getOrNull(1) == "web-group-789")
+
+        // Activity should not crash
+        assert(!activity.isFinishing)
+      }
+    }
+  }
+
+  @Test
```
Comment (vlonfat) [inline app/src/androidTest/java/com/android/joinme/ui/navigation/MainActivityTest.kt]: The test mainActivity_receivesNewIntent_doesNotCrash() do exactly the same thing as the test mainActivity_launchesWithNullDeepLink_doesNotCrash(), please delete the test mainActivity_receivesNewIntent_doesNotCrash()

File: app/src/main/java/com/android/joinme/model/groups/Group.kt [hunk 4]
```diff
@@ -24,6 +24,8 @@ data class Group(
     val ownerId: String = "",
     val memberIds: List<String> = emptyList(),
     val eventIds: List<String> = emptyList(),
-    val membersCount: Int = memberIds.size,
     val photoUrl: String? = null
-)
+) {
```
Comment (vlonfat) [inline app/src/main/java/com/android/joinme/model/groups/Group.kt]: kDoc missing

File: app/src/main/java/com/android/joinme/ui/groups/GroupListScreen.kt [hunk 5]
```diff
@@ -505,3 +609,352 @@ private fun MenuBubble(text: String, icon: ImageVector, onClick: () -> Unit, tes
...
+                  modifier = Modifier.padding(Dimens.Padding.large),
+                  horizontalAlignment = Alignment.CenterHorizontally) {
+                    // Header with close button
+                    Row(
+                        modifier = Modifier.fillMaxWidth(),
+                        horizontalArrangement = Arrangement.SpaceBetween,
+                        verticalAlignment = Alignment.CenterVertically) {
+                          Spacer(Modifier.width(24.dp)) // Balance the close button
+                          Text(
+                              text = "       Join a group",
```
Comment (vlonfat) [inline app/src/main/java/com/android/joinme/ui/groups/GroupListScreen.kt]: Why a big space ?

File: app/src/test/java/com/android/joinme/ui/components/FloatingActionBubblesTest.kt [hunk 6]
```diff
@@ -234,4 +237,359 @@ class FloatingActionBubblesTest {
...
+    }
+
+    // Then: All actions are displayed and clickable
+    composeTestRule.onNodeWithText("Action 1").assertIsDisplayed().assertHasClickAction()
+    composeTestRule.onNodeWithText("Action 2").assertIsDisplayed().assertHasClickAction()
+    composeTestRule.onNodeWithText("Action 3").assertIsDisplayed().assertHasClickAction()
+  }
+
+  @Test
+  fun floatingActionBubbles_firstAction_triggersCallback() {
```
Comment (vlonfat) [inline app/src/test/java/com/android/joinme/ui/components/FloatingActionBubblesTest.kt]: Test floatingActionBubbles_firstAction_triggersCallback() and floatingActionBubbles_lastAction_triggersCallback() must be merged

File: app/src/test/java/com/android/joinme/ui/components/FloatingActionBubblesTest.kt [hunk 7]
```diff
@@ -234,4 +237,359 @@ class FloatingActionBubblesTest {
...
+    composeTestRule.onNodeWithText("Action 3").performClick()
+
+    // Then: Action 3 callback was called
+    assert(action3Clicked) { "Action 3 should have been clicked" }
+  }
+
+  // ==================== Accessibility Tests ====================
+
+  @Test
+  fun floatingActionBubbles_hasClickableActions() {
```
Comment (vlonfat) [inline app/src/test/java/com/android/joinme/ui/components/FloatingActionBubblesTest.kt]: This test does not test anything new because if you can click on it, then it has click action, so it is redundant with the test: floatingActionBubbles_lastAction_triggersCallback().
Please delete it

File: app/src/test/java/com/android/joinme/ui/groups/GroupListScreenTest.kt [hunk 8]
```diff
@@ -1081,4 +1080,1487 @@ class GroupListScreenTest {
     composeTestRule.onNodeWithTag(GroupListScreenTestTags.JOIN_WITH_LINK_BUBBLE).assertExists()
     composeTestRule.onNodeWithTag(GroupListScreenTestTags.CREATE_GROUP_BUBBLE).assertExists()
   }
+
+  // =======================================
+  // Owner/Non-Owner Button Visibility Tests
+  // =======================================
+
+  @Test
+  fun groupMenu_commonButtonsAlwaysVisible() {
```
Comment (vlonfat) [inline app/src/test/java/com/android/joinme/ui/groups/GroupListScreenTest.kt]: This test has no sense now because nobody can have the buttons view group, share group, edit group, leave group, delete group visible at the same time

File: app/src/main/java/com/android/joinme/model/groups/GroupRepositoryFirestore.kt [hunk 9]
```diff
@@ -25,10 +26,11 @@ class GroupRepositoryFirestore(private val db: FirebaseFirestore) : GroupReposit
         Firebase.auth.currentUser?.uid
             ?: throw Exception("GroupRepositoryFirestore: User not logged in.")
 
+    // Get all groups where user is a member (includes groups where they are owner)
     val snapshot =
         db.collection(GROUPS_COLLECTION_PATH).whereArrayContains("memberIds", userId).get().await()
 
-    return snapshot.mapNotNull { documentToGroup(it) }
+    return snapshot.mapNotNull { documentToGroup(it) }.distinctBy { it.id }
```
Comment (vlonfat) [inline app/src/main/java/com/android/joinme/model/groups/GroupRepositoryFirestore.kt]: The distinctBy is not necessary. We fetch all groups from the database where the user is in the userID list. Since a user only appears once in each list (either because they are the owner or because they are a participant), there can be no duplicates.

File: app/src/main/java/com/android/joinme/ui/groups/GroupListViewModel.kt [hunk 10]
```diff
@@ -38,6 +42,10 @@ class GroupListViewModel(
   private val _uiState = MutableStateFlow(GroupListUIState())
   val uiState: StateFlow<GroupListUIState> = _uiState.asStateFlow()
 
+  companion object {
```
Comment (vlonfat) [inline app/src/main/java/com/android/joinme/ui/groups/GroupListViewModel.kt]: You don't need to create a companion object juste a trivial constant. Just put the constant up in the file without  class or object.

**Review by vlonfat — CHANGES_REQUESTED** [review]
# Summary 
Thanks for fixing almost everything in one go!  That's a job well done. However, there are still a couple of things to sort out, which is normal as it's easy to miss a few things. However, the GroupListScreenTest.kt file really needs to be optimized because it is too large. Not only will it add a lot of time to the CI execution, but it will also be complicated to update when changes are made to the GroupListScreen.kt file.

## Remarks
The review was written without AI. The analysis was done again without AI, except for the same file (GroupListScreenTest.kt) because it is still just as large.

## Bugs when using the app
Please delete the android pop up when we want to share a group. It's only a text with not really interesting. It's just annoying because the first pop-up integrated into the application is very good. (All you have to do is delete a few lines and the associated tests.)
Capture d’écran 2025-11-12 à 00 06 58

## Need to be fixed
### GroupListViewModel
- You don't need to create a companion object juste a trivial constant. Just put the constant up in the file without class or object.

### GroupRepositoryFirestore
- The distinctBy is not necessary. We fetch all groups from the database where the user is in the userID list. Since a user only appears once in each list (either because they are the owner or because they are a participant), there can be no duplicates. (getAllGroups)

### GroupListScreenTest.kt
  - The merged test emptyState_displaysCorrectlyWithAllElements already exists. Therefore, the following tests should be removed: emptyState_showsCorrectMessages, emptyState_hasCorrectStyling, and
  emptyState_hasNoListTag (these three tests are redundant with the merged test).
- Merge: fab_showsCorrectIcon_whenClosed with fab_showsCorrectIcon_whenOpen, fabButton_hasAddIcon, and fab_closesJoinBubbles_whenClickedTwice, and then delete them after merging because they will be redundant with the merge test.
- Merge: groupCard_moreButton_isDisplayed with groupCard_moreButton_triggersMenu, groupCard_hasMoreButton, and groupCard_moreButton_hasCorrectIcon, then delete them after merging because they will be redundant with the merge test.
- Merge: shareGroupButton_opensShareDialog with shareGroupDialog_displaysCorrectGroupName,
   shareGroupDialog_displaysShareTitle, shareGroupDialog_displaysHelperText, and shareGroupDialog_displaysDialogTitle, then delete them after merging because they will be redundant with the merge test.
- Merge: leaveGroupButton_notShown_whenOwner with leaveGroupButton_showsCannotLeaveGroup_whenOwner, then delete them after merging because they will be redundant with the merge test. 
- Merge: deleteGroup_buttonNotShown_whenUserIsNotOwner with deleteGroupButton_notShown_whenNotOwner, then delete them after merging as they will be redundant with the merge test.
- Merge: list_isScrollable_and_reaches_last_item with largeList_canScrollToBottom, then delete them after merging because they will be redundant with the merge test.
- Merge: groupCard_withEmptyDescription_doesNotShowDescription with groupWithoutDescription_doesNotShowDescriptionArea and groupCard_blankDescription_notDisplayed, then delete them after merging as they will be redundant with the merged test.
